### PR TITLE
Leetcode solutions

### PR DESCRIPTION
**Pull request by shubhamdpatil**
Leetcode solutions for
2. Add Two Numbers
4. Median of Two Sorted Arrays
7. Reverse Integer
8. String to Integer (atoi)
21. Merge Two Sorted Lists
28. Implement strStr()

#249

### CODE REVIEW

**Comment by danghai** [comment]
Hi @shubhamdpatil , someone already resolved problems 7 and 8. Could you add your solution when you see your way is a different solution? I can merge when you resolve the conflict. Thank you

File: leetcode/src/28.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
+int strStr(char* haystack, char* needle) {
```
Comment (danghai) [inline leetcode/src/28.c]: I think your way is brutal force. Do you come up with more efficient solution? It is a typical problem for pattern search. For example: `Rabin Karp`, `KMP`, or `Horspool`, ...

File: leetcode/src/21.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
+struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
+    struct ListNode *list = NULL;
```
Comment (danghai) [inline leetcode/src/21.c]: Could you add my way below? My way uses recursion.
```c
struct ListNode* mergeTwoLists(struct ListNode* l1, struct ListNode* l2) {
    if(!l1)
        return l2;
    if(!l2)
        return l1;
    if(l1->val < l2->val) {
        l1->next = mergeTwoLists(l1->next, l2);
        return l1;
    } else {
        l2->next = mergeTwoLists(l1, l2->next);
        return l2;
    }
}

```

Comment (shubhamdpatil) [reply leetcode/src/21.c]: Added recusrsive approach to the solution

File: leetcode/src/28.c (hunk 1 above)
Comment (shubhamdpatil) [reply leetcode/src/28.c]: I have implemented this a year ago and I do not want to blindly implement an algorithm, so I'll need some time for that. It is up to you, you can merge this PR and I'll raise another one with the optimised solution to this problem or keep it open till I come up with another approach.

Comment (shubhamdpatil) [reply leetcode/src/28.c]: should I raise a new pull request or add the `kmp` implementation to this pull request.

Comment (danghai) [reply leetcode/src/28.c]: Yes, if you have the chance

Comment (shubhamdpatil) [reply leetcode/src/28.c]: Hi, I could not see my commits in mater branch.

Comment (danghai) [reply leetcode/src/28.c]: You can create a new one

Comment (shubhamdpatil) [reply leetcode/src/28.c]: ok, I'll create a new PR qwith all the previous and will add KMP to it.

**Pull request merged**